File: Go/Day09.go
package main

import AH "./adventhelper"

import (
	"strconv"
	"strings"
)

type Pos struct {
	x, y int
}

type Move struct {
	d string
	n int
}

func parseInput(s string) Move {
	ps := strings.Split(s, " ")
	n, _ := strconv.Atoi(string(ps[1]))
	return Move{ps[0], n}
}

func updateT(posT Pos, posH Pos) Pos {
	dx := posH.x - posT.x
	dy := posH.y - posT.y

	if (AH.AbsInt(dx) > 1) || (AH.AbsInt(dy) > 1) {
		return Pos{posT.x + AH.Sign(dx), posT.y + AH.Sign(dy)}
	} else {
		return posT
	}
}

func pullChain(chain []Pos, histTail map[Pos]int, m Move) ([]Pos) {
		for i := 0; i < m.n; i++ {
		switch m.d {
			case "U":
				chain[0].y -= 1
			case "D":
				chain[0].y += 1
			case "L":
				chain[0].x -= 1
			case "R":
				chain[0].x += 1
		}

		for i := 1; i < len(chain); i++ {
			chain[i] = updateT(chain[i], chain[i-1])
		}

		histTail[chain[len(chain) - 1]] += 1
	}

	return chain
}

func main() {
	lines, _ := AH.ReadStrFile("../input/input09.txt")
	moves := []Move{}
	for _, l := range lines {
		moves = append(moves, parseInput(l))
	}
	hist1 := make(map[Pos]int)
	hist2 := make(map[Pos]int)
	chain1 := []Pos{Pos{0,0}, Pos{0,0}}
	chain2 := []Pos{Pos{0,0}, Pos{0,0}, Pos{0,0}, Pos{0,0}, Pos{0,0}, Pos{0,0},
	               Pos{0,0}, Pos{0,0}, Pos{0,0}, Pos{0,0}}

	for _, m := range moves {
		chain1 = pullChain(chain1, hist1, m)
		chain2 = pullChain(chain2, hist2, m)
	}

	AH.PrintSoln(9, len(hist1), len(hist2))

	return
}

File: C++/Day09.cpp
#include "AH.h"

namespace Day09
{

	typedef std::pair<int, int> Pos;
	typedef std::pair<std::string, int> Move;

	Move ParseInput(const std::string s)
	{
		const auto ps = AH::Split(s, ' ');

		return make_pair(ps[0], std::stoi(ps[1]));
	}

	void update(Pos & t, const Pos h)
	{
		const auto dx = h.first - t.first;
		const auto dy = h.second - t.second;

		if ((std::abs(dx) > 1) || (std::abs(dy) > 1))
		{
			t.first += AH::sgn(dx);
			t.second += AH::sgn(dy);
		}

		return;
	}

	void PullChain(std::vector<Pos> & chain, std::map<Pos, int> & history, const Move m)
	{
		for (int i = 0; i < m.second; ++i)
		{
			if (m.first == "U")
				chain[0].second -= 1;
			else if (m.first == "D")
				chain[0].second += 1;
			else if (m.first == "L")
				chain[0].first -= 1;
			else if (m.first == "R")
				chain[0].first += 1;

			for (size_t j = 1; j < chain.size(); ++j)
				update(chain[j], chain[j - 1]);

			history[chain.back()] += 1;
		}
	}

	int Run(const std::string& filename)
	{
		auto lines = AH::ReadTextFile(filename);
		std::vector<Move> moves;
		for (auto l : lines)
			moves.push_back(ParseInput(l));

		std::map<Pos, int> hist1;
		std::map<Pos, int> hist2;

		std::vector<Pos> chain1;
		for (int i = 0; i < 2; ++i)
			chain1.emplace_back(0 ,0);

		std::vector<Pos> chain2;
		for (int i = 0; i < 10; ++i)
			chain2.emplace_back(0 ,0);

		for (auto m : moves)
		{
			PullChain(chain1, hist1, m);
			PullChain(chain2, hist2, m);
		}

		AH::PrintSoln(9, hist1.size(), hist2.size());

		return 0;
	}

}
